### PR TITLE
20220205 issue18 cmake

### PR DESCRIPTION
**Pull request by Ali-Nasrolahi**
General configuration are almost set and soon be ready to include sources for compiling.

Open to any changes/conflicts. :)

### CODE REVIEW

File: conf/cmake/common.cmake [hunk 1]
```diff
@@ -15,3 +15,71 @@ if(areg MATCHES "static")
...
+    # using Visual Studio C++
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -Werror ${UserDefined}")
+endif()
+
+
+
+if(Config MATCHES "Release")
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -O2 -DNDEBUG")
+else()
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -DNDEBUG")
```
Comment (biwiki) [inline conf/cmake/common.cmake]: I think that it should be -DDEBUG here for Debug builds

File: conf/cmake/user.cmake [hunk 2]
```diff
@@ -38,16 +40,16 @@ set(UserDefLibPaths)
 
 set(UserDefLibs)
 
-set(UserDefOutput)
+set(UserDefOutput "product")
```
Comment (aregtech) [inline conf/cmake/user.cmake:46]: @Ali-Nasrolahi , there are too many new lines in the file :)
Right now it is OK, but please do not forget at the end to make some cleanups.
Thank you!

File: examples/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,9 @@
+if($(AREG_BINARY) MATCHES "shared")
+    #These variables must be included in CMAKE_CXX_FLAGS
+	set(examples_CXXFLAGS "-DIMP_AREG_DLL")
+	set(examples_LDFLAGS  "Wl,-rpath=${AREG_OUTPUT_BIN} -L ${AREG_OUTPUT_BIN} -lareg ${LDFLAGS}")
+else()
+    #These variables must be included in CMAKE_CXX_FLAGS
+	set(examples_CXXFLAGS "-DIMP_AREG_LIB")
+	set(examples_LDFLAGS  "-L ${AREG_OUTPUT_LIB} -Wl,-Bstatic -lareg -Wl,-Bdynamic ${LDFLAGS}")
+endif()
```
Comment (aregtech) [inline examples/CMakeLists.txt]: @Ali-Nasrolahi , please add new line at the and of the file.

File: framework/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,2 @@
+include("${AREG_BASE}/areg/CMakeLists.txt")
+include("${AREG_BASE}/mcrouter/CMakeLists.txt")
```
Comment (aregtech) [inline framework/CMakeLists.txt]: @Ali-Nasrolahi , please add new line at the and of the file.

**Review by aregtech — APPROVED** [review]
@Ali-Nasrolahi ,
thank you for your changes! There are very minor things to fix. Can you please fix them?

File: conf/cmake/user.cmake (hunk 2 above)
Comment (Ali-Nasrolahi) [reply conf/cmake/user.cmake:46]: Sure no problem at all.

File: conf/cmake/common.cmake (hunk 1 above)
Comment (Ali-Nasrolahi) [reply conf/cmake/common.cmake]: Thanks for noticing. :)

File: conf/cmake/common.cmake [hunk 5]
```diff
@@ -38,9 +38,9 @@ endif()
 
 
 if(Config MATCHES "Release")
-    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -O2 -DNDEBUG")
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -O2 -DDEBUG")
```
Comment (biwiki) [inline conf/cmake/common.cmake:41]: Nah, for Release builds we use -DNDEBUG :-), but this can be addressed later

**Review by biwiki — APPROVED** [review]
It is so closed to have a new build system

File: conf/cmake/common.cmake (hunk 5 above)
Comment (aregtech) [reply conf/cmake/common.cmake:41]: @Ali-Nasrolahi , I merge these changes, but please keep in mind that for release build we use "-DNDEBUG" option.
-DDEBUG or -D_DEBUG is for Debug build
-DNDEBUG or -D_NDEBUG for Release build
Maybe all compiler options I document in one place :) There are not too many.

**Review by aregtech — APPROVED** [review]
@Ali-Nasrolahi , thank you for your changes!

**Pull request merged**